python/ray/serve/_private/proxy_state.py

```diff
@@ -30,35 +33,282 @@
+            if finished:
+                worker_id, log_file_path = json.loads(ray.get(finished[0]))
+                self.worker_id = worker_id
+                self.log_file_path = log_file_path
+                return ActorWrapperCallStatus.FINISHED_SUCCEED
+            else:
+                return ActorWrapperCallStatus.PENDING
+        except RayActorError:
+            return ActorWrapperCallStatus.FINISHED_FAILED
+
+    def is_healthy(self) -> ActorWrapperCallStatus:
+        """Return whether the proxy actor is healthy or not.
+
+        If the ongoing health check is finished, and the value can be retrieved,
+        reset _health_check_obj_ref to enable the next health check and return
+        FINISHED_SUCCEED status. If the ongoing ready check is not finished,
+        return PENDING status. If the RayActorError is raised, meaning that the actor
+        is dead, return FINISHED_FAILED status.
+        """
+        try:
+            finished, _ = ray.wait([self._health_check_obj_ref], timeout=0)
+            if finished:
+                self._health_check_obj_ref = None
+                ray.get(finished[0])
+                return ActorWrapperCallStatus.FINISHED_SUCCEED
+            else:
+                return ActorWrapperCallStatus.PENDING
+        except RayActorError:
+            return ActorWrapperCallStatus.FINISHED_FAILED
+
+    def is_drained(self) -> ActorWrapperCallStatus:
+        """Return whether the proxy actor is drained or not.
+
+        If the ongoing drained check is finished, and the value can be retrieved,
+        reset _is_drained_obj_ref to ensure drained check is finished and return
+        FINISHED_SUCCEED status. If the ongoing ready check is not finished,
+        return PENDING status.
+        """
+        finished, _ = ray.wait([self._is_drained_obj_ref], timeout=0)
+        if finished:
+            self._is_drained_obj_ref = None
+            ray.get(finished[0])
+            return ActorWrapperCallStatus.FINISHED_SUCCEED
+        else:
+            return ActorWrapperCallStatus.PENDING
+
+    def is_shutdown(self) -> bool:
+        """Return whether the proxy actor is shutdown.
+
+        If the actor is dead, the health check will return RayActorError.
+        """
+        try:
+            ray.get(self._actor_handle.check_health.remote(), timeout=0)
+        except RayActorError:
+            # The actor is dead, so it's ready for shutdown.
+            return True
+
+        # The actor is still alive, so it's not ready for shutdown.
+        return False
+
+    def update_draining(self, draining: bool):
+        """Update the draining status of the proxy actor."""
+        self._update_draining_obj_ref = self._actor_handle.update_draining.remote(
+            draining, _after=self._update_draining_obj_ref
+        )
+
+    def kill(self):
+        """Kill the proxy actor."""
+        ray.kill(self._actor_handle, no_restart=True)
+
+    @classmethod
+    def start_proxy(
```

Comment:
nit: move this to the top after the constructor (standard for classmethod constructors)

Reply:
also, should this just be the normal constructor instead of a class method? don't see a need for the way it's currently structured

Reply:
oh good point, I will just merge this into `__init__`